### PR TITLE
Fixed smart count false positive on subquery JOINs

### PR DESCRIPTION
**Pull request by kevinansfield**
## Summary
- `hasMultiTableSource` matched `/\bjoin\b/i` against the fully compiled SQL, so a `where id in (select … inner join …)` forced `count(distinct posts.id)` even when the outer query was still single-table. On a 267k-post Ghost site filtered by author this was a measured ~20% regression on the count aggregate (316ms vs 257ms) for zero correctness benefit — the weedout step already guaranteed distinct rows.
- Switched detection to walk knex's query-builder AST (`_statements` + `_single.table`) instead of regex-matching compiled SQL. Joins, unions, derived tables and comma-separated FROM lists are all read as structured data, and JOINs nested in `WHERE id IN (…)` subqueries live under a `where` grouping and are correctly ignored.
- Since bookshelf is built on knex, the builder always exposes `_statements` / `_single` — no SQL fallback needed. The whole check collapses to a ~30-line AST walk: no regexes, no `toSQL()` call, no paren-stripping, no derived helpers.

## Tests
Added `bookshelf@1.2.0`, `knex@3.2.9` and `sqlite3@5.1.7` as devDependencies so the plugin can be tested against real knex/bookshelf instances instead of hand-rolled stubs. 25 tests total, 100% line/branch/function coverage on `lib/bookshelf-pagination.js`.

- **14 stub tests** (`test/pagination.test.js`) — pagination metadata math, `limit='all'`, `transacting` passthrough, error wrapping (NotFound / BadRequest / unknown), and the `parseOptions` / `formatResponse` utilities. None of these need a real DB.
- **9 direct `hasMultiTableSource` tests** (`test/pagination-integration.test.js`) — call the function against real knex QueryBuilders (no DB connection) and cover every branch: plain single-table, subquery-in-WHERE, innerJoin, leftJoin/joinRaw, UNION, derived table in FROM, `fromRaw` with multiple tables, comma-string table, and CTE-only. These pin the plugin's assumptions about knex's AST shape to the version of knex actually installed, and cover edge cases that can't be driven through `fetchPage` end-to-end (UNION + aggregate is semantically weird; a comma-string table fails at sqlite execution).
- **2 end-to-end `fetchPage` tests** — the only two scenarios that carry unique weight on top of the direct suite:
  1. An inner join that duplicates rows (p1 × two tags → 3 physical rows for 2 published posts). This is the only shape where `count(*)` and `count(distinct)` actually produce different totals, so it's the only test that catches a "we picked the wrong aggregate" bug at the result level.
  2. A subquery JOIN in WHERE — the named regression this PR is about, kept as an explicit end-to-end guard that the fix reaches production via `fetchPage`.

Each e2e test listens on knex's `query` event to assert *which* count aggregate the plugin chose, and asserts `pagination.total` against a value that's only correct if that aggregate was picked — so "we picked count(*)" and "the result is right" are verified together.

## Notes on upstream consumers
This is strictly safe for anyone already using `useSmartCount: true`: every case the old code flagged as multi-table is still flagged by the AST walk. The change only *removes* false positives (subquery JOINs) and *adds* one previously-missed case (derived tables in FROM). `useBasicCount` and the default distinct-count behaviour are untouched.

### CODE REVIEW

File: packages/bookshelf-pagination/lib/bookshelf-pagination.js [hunk 1]
```diff
@@ -14,46 +14,40 @@ const messages = {
...
-// 2) old-style comma-separated FROM lists (eg `from posts, tags`)
-// Both can duplicate base rows, so fetchPage should use count(distinct id).
-function hasMultiTableSource(queryBuilder) {
-    const sql = getCompiledSql(queryBuilder);
+    // For a bookshelf Model the outer FROM is normally a plain table name
+    // string. Anything else — a QueryBuilder (derived table) or a Raw (eg.
+    // `fromRaw('a, b')`) — can carry duplicate-producing shape that we can't
+    // introspect further, so treat it as multi-source.
+    const table = queryBuilder._single.table;
+    if (table && typeof table !== 'string') {
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply packages/bookshelf-pagination/lib/bookshelf-pagination.js:43]: False positive — not applying.

Why the finding doesn't hold:

1. Internal caller (lib/bookshelf-pagination.js:231): hasMultiTableSource(countQuery) is invoked with this.query().clone() — a bookshelf Model's underlying knex QueryBuilder. Every knex builder has _statements: [] and _single: {} initialised at construction; .clone() preserves them.
2. External caller: hasMultiTableSource is exposed on paginationUtils only so the tests can exercise it against real knex builders. Nothing else in the monorepo calls it (confirmed by grep).
3. Empirical check: test/pagination-integration.test.js runs hasMultiTableSource directly against 9 real knex builder shapes (plain table, subquery-in-WHERE, innerJoin, leftJoin/joinRaw, UNION, derived table, fromRaw, comma-string table, CTE) plus 8 full fetchPage flows against real bookshelf. If _statements or _single could be missing, a TypeError would surface there. All 31 tests pass.
4. The suggested fallback direction is wrong in substance: the reviewer wants "missing internals → treat as multi-table → use count(distinct)". But if internals were ever actually missing, that behavior would silently undo the useSmartCount optimization on every call — reintroducing the exact regression this PR fixes. A loud TypeError during development is strictly better than a silent 20% perf regression in production.
5. Prior review history: the earlier code-quality review pass in this session explicitly flagged the old if (!Array.isArray(statements) && (!single || typeof single !== 'object')) guard as dead defensiveness, and we removed it for the same reason.
